fix(server): use fqdn for og:image meta tag value

eleith opened the pull request:
opengraph image specifies that the url contains http or https, thus implying a fqdn.

this change uses the external domain from the server config to attempt to make the og:image have both the existing path to the thumbnail along with the desired domain

please note, some og implementations do work with relative paths, so not all og image checkers may still pass, but not all implementations have this fallback and thus will not find the image otherwise

fix #11081

Commit:
attempt to use fqdn for og:image

opengraph image specifies that the url contains http or https, thus
implying a fqdn.

this change uses the external domain from the server config to attempt
to make the og:image have both the existing path to the thumbnail along
with the desired domain

if the server setting is empty, the old behavior will persist

please note, some og implementations do work with relative paths, so not
all og image checkers may still pass, but not all implementations have
this fallback and thus will not find the image otherwise

mertalev commented on web/src/routes/(user)/share/[key]/[[photos=photos]]/[[assetId=id]]/+page.ts:
```diff
@@ -2,19 +2,23 @@ import { getAssetThumbnailUrl, setSharedLink } from '$lib/utils';
 export const load = (async ({ params }) => {
   const { key } = params;
   await authenticate({ public: true });
   const asset = await getAssetInfoFromParam(params);
+  const config = await getConfig()
+  const domain = config.server.externalDomain;
```
You can move these two in the try/catch and put `getAssetInfoFromParam`, `getConfig` and `getMySharedLink` in a `Promise.all` call. It's better to do this in parallel instead of waiting for each to finish.

jrasm91 reviewed (dismissed):
This is a bit pointless since the change isn't server side rendered. You should take a look at api.service.ts in the server repository instead. There is also an e2e test for shared links that checks this specific functionality.

Commit:
tests and ssr for og:image value as fqdn

Commit:
formatting

Commit:
fix test

Commit:
formatting

Commit:
formatting

Commit:
fix tests

getConfig was requiring authentication. using already initiated global stores instead

eleith commented:
i've addressed the comments, lints and tests. there is likely some opportunity to improve the approach (SSR vs web) to improve future DX concerns, but they could be decoupled from this PR as well. i am open to either. 

thanks for the feedback so far.

jrasm91 commented on server/src/services/api.service.ts:
```diff
@@ -69,13 +75,14 @@ export class ApiService {
         return next();
       }
 
+      const config = await this.configCore.getConfig({ withCache: true });
```
The code that loads the config should be inside of the getMetadataTags method in the shared link service.

eleith replied:
implemented

jrasm91 reviewed (commented):
Looks a lot better.

Commit:
load config in shared link service itself

jrasm91 reviewed (approved):
This looks good, the only issue iM thinking about now is if the extension domain has a trailing slash or not? Did you test both situations? Alternatively new URL(path, domain).href would probably fix the issue as well

Commit:
join host and pathname/params safely

eleith commented:
> This looks good, the only issue iM thinking about now is if the extension domain has a trailing slash or not? Did you test both situations? Alternatively new URL(path, domain).href would probably fix the issue as well

i didn't test that! 

i just dove in assuming `externalDomain` was a trimmed and validated host and lazily joined them. 

i've pushed updates now to better assemble the host and pathname/query params and have successfully tested with trailing slashes.

jrasm91 commented:
I mean, we probably should validate/normalize it lol. Maybe in a follow up PR we'll fix it all?

eleith commented:
I'll fix the tests

Commit:
use origin instead of host for full domain string

also fixes lint and address the imageURL type which is optional

eleith commented:
locally, lint, e2e and svelte check now all pass for me

eleith commented:
@jrasm91 I assume I don't have to keep this PR updated until further notice.

let me know if there are additional requirements or actions I can take.

jrasm91 commented:
> @jrasm91 I assume I don't have to keep this PR updated until further notice.
> 
> let me know if there are additional requirements or actions I can take.

Correct. It has just been a busy week, but I will try to get it merged in next week.

eleith commented:
thanks for all the feedback and i appreciate understanding the next steps.

I mostly wanted clarification that I don't need to monitor the PR anymore.

all sounds good to me, it will land when it lands!

jrasm91 committed:
chore: clean up
